packages/flutter_tools/lib/src/base/file_system.dart

```diff
@@ -170,6 +129,50 @@ class FileSystemUtils {
   }
 }
 
+/// Creates `destDir` if needed, then recursively copies `srcDir` to
+/// `destDir`, invoking [onFileCopied], if specified, for each
+/// source/destination file pair.
+///
+/// Skips files if [shouldCopyFile] returns `false`.
+void copyDirectory(
+  Directory srcDir,
+  Directory destDir, {
+  bool shouldCopyFile(File srcFile, File destFile),
+  void onFileCopied(File srcFile, File destFile),
+}) {
+  if (!srcDir.existsSync()) {
+    throw Exception('Source directory "${srcDir.path}" does not exist, nothing to copy');
+  }
+
+  if (!destDir.existsSync()) {
+    destDir.createSync(recursive: true);
+  }
+
+  for (final FileSystemEntity entity in srcDir.listSync()) {
+    final String newPath = destDir.fileSystem.path.join(destDir.path, entity.basename);
+    if (entity is Link) {
```

Comment:
Example of a link falling through to the exception: https://github.com/flutter/flutter/issues/35170